### DISCUSSION TITLE
Post Event after orchestrator complete

### DISCUSSION THREAD

**Discussion by tech7857** [Azure/azure-functions-durable-extension#2237] [Q&A]
Hi 

we want to send an event to an external system when an orchestration completes i.e marked as complete or failed . I couldnt find any trigger that I can put on the storage table.

can someone please suggest on how we can achieve this.

thanks

**Comment by davidmrdavid** [comment]
Hi @tech7857:

I see you also posted on the Python SDK repo about this question. In general, I don't think we have a trigger that checks specifically for whether an orchestrator has completed, but you could use the `sub-orchestrator` pattern described here (https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-sub-orchestrations?tabs=csharp) to build something like what you're asking for.

Please let me know if that works for you.

**Reply by tech7857** [reply]
Hi @davidmrdavid 

can you please some more guidance on how to achieve this through sub-orchestrator.

it is not very clear on where to add the event when orchestrator completes

**Reply by davidmrdavid** [reply]
Hi @tech7857.

Sure, please consider the following sample (in Python) - https://github.com/Azure/azure-functions-durable-python/tree/dev/samples/simple_sub_orchestration

Here we have 2 orchestrators - one called `DurableSubOrchestrator` and another called `DurableOrchestrator`. Essentially, `DurableOrchestrator` invokes `DurableSubOrchestrator` via the `call_suborchestrator` API. That this does, is that it makes `DurableOrchestrator` _wait_ until `DurableSubOrchestrator` returns, and it then receives the output value of `DurableSubOrchestrator` as the result of `call_suborchestrator`.

So what you can do is that, underneath the `call_suborchestrator` API call, you can write whatever logic you need to call after you receive your orchestrator result.

Does that make sense?

**Reply by tech7857** [reply]
Hi @davidmrdavid 

thanks for the input and taking time to help. Apologies for my lack of understanding but even if the sub orchestrator returns and I add a logic to send an event but the main orchestrator will still not be complete and there can be a time lag in getting the main orchestrator to complete and when the consumer calls the status uri of the orchestrator to get the results.

please correct if I am missing anything here.

**Reply by davidmrdavid** [reply]
Hi @tech7857, no worries at all - this stuff is complicated :-)

So my suggestion was that your parent orchestrator would invoke your "consumer" directly. You could do this by writing some code like this in your parent orchestrator:

```Python
def orchestrator_function(context: df.DurableOrchestrationContext):
    result = yield context.call_sub_orchestrator("DurableSubOrchestrator", "Seattle")
    yield context.call_activity("CallMyConsumer", result)
    return None # you don't really need the result of this orchestrator
```

In essence, you can notify your consumer that the sub-orchestrator has completed via an Activity invocation in the parent.
Does that fit your scenario? Perhaps I'm misunderstanding your question. Please let me know!

**Reply by tech7857** [reply]
Hi @davidmrdavid 

Thanks again. Our use case is we are trying to implement an HTTP Async pattern where the consumer will call an HTTP start endpoint which will trigger the async process and once the process is complete, will send a notification to the consumer that the process is complete and they can call another endpoint (status endpoint(client.get_status)) passing the orchestrator instance id and get there results. What we are trying to solve  is to send the event to the consumer when the orchestrator has the completed state, so that when they call the status end point they get the result back. We dont want them to poll the status endpoint continuously to check if the results are available or not.

Currently the problem is we are sending that event from the activity function and sometimes when they call the status endpoint the orchestrator status is still running, which defeats the purpose.

We were trying to see if there an trigger which we can right on storage table, so that if status is complete, we can trigger an azure function which will send this event and it is guaranteed whenever consumer will call us the result will always be available.

Please let me know if this helps. Also, I will try out what you provided but I am still unsure will this really solve the problem.

Thanks

**Reply by davidmrdavid** [reply]
Thanks @tech7857.  Have you looked into this documentation: https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-event-publishing?tabs=csharp-script

It describes a means of receiving lifecycle notifications for monitors, which may be used for the kind of notification system you're thinking about. Please give it a look and let me know if this fits your usecase.
